File: python/lesson-08-fastapi/exercises/exercise.py
"""Lesson 08 Exercise — Build a Product CRUD API with FastAPI"""

from fastapi import FastAPI, HTTPException, status

app = FastAPI(title="Product API", version="1.0.0")

# TODO: Define Pydantic models:
#   - CreateProductRequest(name: str, price: float)
#       - validate: name not empty, price >= 0
#   - UpdateProductRequest(name: str | None, price: float | None)
#       - validate: at least one field provided
#   - ProductResponse(id: int, name: str, price: float)

# TODO: Create a ProductService class with an in-memory dict store
#   seeded with 3 products. Implement:
#   - get_all() -> list[ProductResponse]
#   - get_by_id(id) -> ProductResponse | None
#   - create(req) -> ProductResponse
#   - update(id, req) -> ProductResponse | None
#   - delete(id) -> bool

# TODO: Wire up routes:
#   GET    /api/products         → {count, data}
#   GET    /api/products/{id}    → product or 404
#   POST   /api/products         → 201 + created product
#   PUT    /api/products/{id}    → updated product or 404
#   DELETE /api/products/{id}    → message or 404

# Run with: uvicorn exercise:app --reload --port 8001

File: dotnet/lesson-08-http-api/exercises/Exercise.cs
// ── Lesson 08 Exercises ───────────────────────────────────────────────────────
// Build a minimal Product CRUD API.

var builder = WebApplication.CreateBuilder(args);
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

// TODO: Register a singleton IProductService (in-memory implementation)

var app = builder.Build();
app.UseSwagger();
app.UseSwaggerUI();

// TODO: Map GET    /api/products         → return all products with count
// TODO: Map GET    /api/products/{id}    → return product or 404
// TODO: Map POST   /api/products         → create (name + price required), return 201
// TODO: Map PUT    /api/products/{id}    → update name/price (partial), return 200 or 404
// TODO: Map DELETE /api/products/{id}    → delete, return 200 or 404

app.Run("http://localhost:5001");

// ── Models ────────────────────────────────────────────────────────────────────

// TODO: Define a Product record (Id, Name, Price)
// TODO: Define CreateProductRequest and UpdateProductRequest records

// ── Service ───────────────────────────────────────────────────────────────────

// TODO: Define IProductService interface
// TODO: Implement InMemoryProductService seeded with 3 products
